src/main/java/edu/harvard/iq/dataverse/api/Admin.java

```diff
@@ -1920,4 +1922,103 @@ public Response getBannerMessages(@PathParam("id") Long id) throws WrappedRespon
 
     }
     
+    @GET
+    @Path("/licenses")
+    public Response getLicenses() {
+    	JsonArrayBuilder arrayBuilder = Json.createArrayBuilder();
+		for(License license : licenseService.listAll()) {
+			try {
+				arrayBuilder.add(JsonPrinter.json(license));
+			} catch (URISyntaxException e) {
+				return error(Status.INTERNAL_SERVER_ERROR, "Incorrect URI in JSON");
+			}
+		}
+        return ok(arrayBuilder);
+    }
+
+    @GET
+    @Path("/licenses/id/{id}")
+    public Response getLicenseById(@PathParam("id") long id) {
+		try {
+			License license = licenseService.getById(id);
+        	return ok(json(license));
+		} catch (FetchException e) {
+			return error(Response.Status.NOT_FOUND, e.getMessage());
+		} catch (URISyntaxException e) {
+			return error(Response.Status.BAD_REQUEST, "Incorrect URI in JSON");
+		}
+	}
+
+    @GET
+    @Path("/licenses/name/{name}")
+    public Response getLicenseByName(@PathParam("name") String name) {
+		try {
+			License license = licenseService.getByName(name);
+        	return ok(json(license));
+		} catch (FetchException e) {
+			return error(Response.Status.NOT_FOUND, e.getMessage());
+		} catch (URISyntaxException e) {
+			return error(Response.Status.BAD_REQUEST, "Incorrect URI in JSON");
+		}
+	}
+
+    @POST
+    @Path("/licenses")
+    public Response addLicense(License license) {
+        try {
+            licenseService.save(license);
+            return created("/api/admin/licenses", Json.createObjectBuilder().add("message", "License created"));
+        } catch (RequestBodyException e) {
+			return error(Response.Status.BAD_REQUEST, e.getMessage());
+		} catch(PersistenceException e) {
+            return error(Response.Status.CONFLICT, "A license with the same URI or name is already present.");
+        }
+	}
+
+    @PUT
+    @Path("/licenses/id/{id}")
+    public Response putLicenseById(@PathParam("id") long id, License license) {
+		try {
+			licenseService.setById(id, license.getName(), license.getShortDescription(), license.getUri(), license.getIconUrl(), license.isActive());
+		} catch (UpdateException e) {
+			return error(Response.Status.BAD_REQUEST, e.getMessage());
```

Comment:
Here you also return the original message with e.getmessage(), this might give more information useful to the client. Could we do this for all error's, so we know the URI or name that is problematic, for instance.

Reply:
The admin won't get to see more info in either case. The responses with e.getMessage() basically return messages that were thrown lower down the chain, whereas the ones with a String arg come from that same level. They both return custom messages and no other info.